docs/features/quartz plugin.md

```diff
@@ -0,0 +1,40 @@
+---
+title: 📦 quartz-plugin
+---
+
+> [!info]
+>
+> `quartz-plugin` allows you to manage plugins built by Quartz community members.
+
+On this page, "shorthand" refers to the syntax `author/plugin_name`, which will expand to `https://github.com/author/plugin_name.git`.
+
+"name" would simply be `plugin_name`.
+
+## Command reference
+
+| Command   | Arguments               | Effect                                                                      | Shorthand?   |
+| --------- | ----------------------- | --------------------------------------------------------------------------- | ------------ |
+| `add`     | `url`: required         | Adds the supplied plugin to `quartz.plugins.ts` with default configuration. | Yes          |
+| `remove`  | `url` \| name: required | Removes a plugin's files and configuration                                  | Yes          |
+| `enable`  | $\downarrow$            | $\downarrow$                                                                | $\downarrow$ |
+| `disable` | $\downarrow$            | $\downarrow$                                                                | $\downarrow$ |
+| `toggle`  | `url` \| name: required | Sets `enable:` to `true`, `false`, or `!enabled`, respectively.             | Yes          |
```

Comment:
preferably we should do this in config, rather than CLI.

Otherwise this would be way too convoluted and users won't know where to start.

also make quartz managing configs and state very hard and even more complex.

I'm ok with `add`, `delete`, `update` ops.

Reply:
It is in config, `enable` is part of the filter I apply and it's what would be changed by those three. It's less work for some workloads to just bang out a terminal command instead of opening the file, but not implementing it saves me some work so that's fine haha

Reply:
sometimes implementing something nice would incur a lot of overhead for maintaining logic. You are essentially will have to figure out the order of the plugins and managing state between load.

Reply:
You've mentioned load order and cross-build twice now, but bo my knowledge, neither of those concepts is in Quartz at the moment; it just looks for a change in config and hard builds when doing so in the context's plugin array order. Plugin developers configure when they want their plugin to run in two ways:
- plugin type: filters exclude files pre-build, emitters run at build time, and transformers run at build time and push scripts for the client. 
- load time: the DOM is a pretty significant inflection point. Any external resources pushed by plugins can specify their content type and load time in terms of DOM events. Again, this is on plugin devs, not the user. 
  - and to the extent that plugins need to require other plugins at build time, our bundler handles all dependency resolution. And on client side you'll have to start getting into async-await because there are no order guarantees when this hits a browser.

Reply:
But the order of the transformers is also important as well afaik. Maybe jacky can correct me on this.

In terms of DOM event, it is just `afterDOMLoaded` and `beforeDOMLoaded` right? the bundler doesn't need to now any runtime js.

On client side I think plugins should be ready even before hitting the browser no?

Or am I missing something here?

Reply:
If the order of the transformer objects currently in config ts is significant, then that's a Quartz problem and I'll open an issue to request a Quartz backend plugin DAG. 

Re: *DomLoaded, client scripts are just the only other case I can think of where order matters. That event solves the problem that script tags on your webpage have no guaranteed load order. If a dev does need to run two script tags on the same dom event but in a specific order, that's outside the purview of the CLI plugin installation and config manager and the dev should look to signals, blocking, etc. to ensure the order.

Reply:
FYI https://quartz.jzhao.xyz/advanced/creating-components

All script is either loaded before DOM or after DOM in quartz system.

I do think we need a DAG for plugin system. Probably we should do that first 🤔

Reply:
imo no need for a DAG if we linearize it anyways at parse time lol the whole https://github.com/unifiedjs/unified ecosystem is a pipe, can't jam a DAG in there

Reply:
> imo no need for a DAG if we linearize it anyways at parse time lol the whole https://github.com/unifiedjs/unified ecosystem is a pipe, can't jam a DAG in there

Do you have any input on the mentioned situation where one transformer needs to run before `ofm` etc?

Reply:
yes e.g. TOC needs to run after latex and ofm so that the header links have extra syntax parsed, we get some performance improvements running latex before syntax highlighting so shikiji doesn't also highlight the math as a codeblock

i have thoughts that I'm writing up at a high-level rn gimme a sec

Reply:
![quartz.config.ts, plugins, transformers](https://github.com/jackyzha0/quartz/assets/138795235/5b682afc-3d1e-4e2f-9342-a5f127c8e918)
so currently if someone decided to alphabetize their transformers list it would break quartz?

that's cursed lmao
...+1 for dag in `build.ts`

Reply:
Ye you can try move syntaxHighlighting above latex, or toc before latex it won't work as expected.

Reply:
strong disagree lol we should not be adding a DAG to the build process here, especially with community plugins you can get weird cases that cause a complete topology breakdown with poorly written rules

dag does not buy us anything here if we need to flatten it anyways for unified

Reply:
dependency graphs scare people